# Overview
- Load statistical metrics from EDF files and transform into a dataframe
- Observe dataframe dimension and data types
- Save dataframe to csv file

In [10]:
# Import the necessary packages from utils file
from utils import pd, np, mne, os

In [11]:
# Load one EDF file (PhysioNet uses .edf format)
path_name = '../data/raw/'

# Get all EDF files from the raw directory
edf_files = [f for f in os.listdir(path_name) if f.endswith('.edf')]

# Create a table of features
features = []
labels = []

# Create column names for features
feature_names = ['trial_id','subject_id']
for ch in range(1, 65):  # 64 channels
    feature_names.extend([
        f'ch{ch}_std',
        f'ch{ch}_mean',
        f'ch{ch}_max',
        f'ch{ch}_min'
    ])

# Process each file
raw_data = []
for file_name in edf_files:
    raw = mne.io.read_raw_edf(path_name + file_name, preload=True, encoding='latin1')
    raw.filter(1., 40.)  # Bandpass filtering
    raw_data.append(raw)

    # Assume annotations are already present
    events, event_id = mne.events_from_annotations(raw)

    # Epoching (trial segmentation, e.g., 0–1s after event)
    epochs = mne.Epochs(raw, events, event_id=event_id, tmin=0, tmax=1, baseline=None, preload=True)

    # Extract trial IDs (sample index of each event)
    trial_ids = epochs.events[:, 0]  # Sample index of event onset

    for trial_id, epoch, label in zip(trial_ids, epochs.get_data(), epochs.events[:, -1]):
    # epoch shape: (n_channels, n_times)
        channel_features = [trial_id, file_name.split('.')[0]]  # Extract subject ID from filename
        for channel in epoch:
            channel_features.extend([
                np.std(channel),
                np.mean(channel),
                np.max(channel),
                np.min(channel)
            ])
        features.append(channel_features)
        labels.append(label)

Extracting EDF parameters from /Users/miriamlandau/Documents/predict_hand_imagery/data/raw/S012R04.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19679  =      0.000 ...   122.994 secs...
Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 1 - 40 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 1.00
- Lower transition bandwidth: 1.00 Hz (-6 dB cutoff frequency: 0.50 Hz)
- Upper passband edge: 40.00 Hz
- Upper transition bandwidth: 10.00 Hz (-6 dB cutoff frequency: 45.00 Hz)
- Filter length: 529 samples (3.306 s)

Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
Not setting metadata
30 matching events found
No baseline correction applied
0 projection items activated
Using data 

[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished


Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
Not setting metadata
30 matching events found
No baseline correction applied
0 projection items activated
Using data from preloaded Raw for 30 events and 161 original time points ...
0 bad epochs dropped
Extracting EDF parameters from /Users/miriamlandau/Documents/predict_hand_imagery/data/raw/S006R04.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19679  =      0.000 ...   122.994 secs...
Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 1 - 40 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 1.00
- Lower transition bandwidth: 1.00 Hz (-6 dB cutoff frequency: 0.50 Hz)
- Upper passband edge: 40.00 Hz
- Upper transition bandwidt

[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.1s finished
[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished


Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
Not setting metadata
30 matching events found
No baseline correction applied
0 projection items activated
Using data from preloaded Raw for 30 events and 161 original time points ...
0 bad epochs dropped
Extracting EDF parameters from /Users/miriamlandau/Documents/predict_hand_imagery/data/raw/S008R04.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19679  =      0.000 ...   122.994 secs...
Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 1 - 40 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 1.00
- Lower transition bandwidth: 1.00 Hz (-6 dB cutoff frequency: 0.50 Hz)
- Upper passband edge: 40.00 Hz
- Upper transition bandwidt

[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished


Extracting EDF parameters from /Users/miriamlandau/Documents/predict_hand_imagery/data/raw/S013R04.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19679  =      0.000 ...   122.994 secs...
Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 1 - 40 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 1.00
- Lower transition bandwidth: 1.00 Hz (-6 dB cutoff frequency: 0.50 Hz)
- Upper passband edge: 40.00 Hz
- Upper transition bandwidth: 10.00 Hz (-6 dB cutoff frequency: 45.00 Hz)
- Filter length: 529 samples (3.306 s)

Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
Not setting metadata
30 matching events found
No baseline correction applied
0 projection items activated
Using data 

[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished


Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
Not setting metadata
30 matching events found
No baseline correction applied
0 projection items activated
Using data from preloaded Raw for 30 events and 161 original time points ...
0 bad epochs dropped
Extracting EDF parameters from /Users/miriamlandau/Documents/predict_hand_imagery/data/raw/S011R04.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19679  =      0.000 ...   122.994 secs...
Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 1 - 40 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 1.00
- Lower transition bandwidth: 1.00 Hz (-6 dB cutoff frequency: 0.50 Hz)
- Upper passband edge: 40.00 Hz
- Upper transition bandwidt

[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.1s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.1s finished
[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished


Setting up band-pass filter from 1 - 40 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 1.00
- Lower transition bandwidth: 1.00 Hz (-6 dB cutoff frequency: 0.50 Hz)
- Upper passband edge: 40.00 Hz
- Upper transition bandwidth: 10.00 Hz (-6 dB cutoff frequency: 45.00 Hz)
- Filter length: 529 samples (3.306 s)

Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
Not setting metadata
30 matching events found
No baseline correction applied
0 projection items activated
Using data from preloaded Raw for 30 events and 161 original time points ...
0 bad epochs dropped
Extracting EDF parameters from /Users/miriamlandau/Documents/predict_hand_imagery/data/raw/S007R04.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ..

[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.2s finished


Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
Not setting metadata
30 matching events found
No baseline correction applied
0 projection items activated
Using data from preloaded Raw for 30 events and 161 original time points ...
0 bad epochs dropped
Extracting EDF parameters from /Users/miriamlandau/Documents/predict_hand_imagery/data/raw/S005R04.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 19679  =      0.000 ...   122.994 secs...
Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 1 - 40 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 1.00
- Lower transition bandwidth: 1.00 Hz (-6 dB cutoff frequency: 0.50 Hz)
- Upper passband edge: 40.00 Hz
- Upper transition bandwidt

[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.0s finished


Filtering raw data in 1 contiguous segment
Setting up band-pass filter from 1 - 40 Hz

FIR filter parameters
---------------------
Designing a one-pass, zero-phase, non-causal bandpass filter:
- Windowed time-domain design (firwin) method
- Hamming window with 0.0194 passband ripple and 53 dB stopband attenuation
- Lower passband edge: 1.00
- Lower transition bandwidth: 1.00 Hz (-6 dB cutoff frequency: 0.50 Hz)
- Upper passband edge: 40.00 Hz
- Upper transition bandwidth: 10.00 Hz (-6 dB cutoff frequency: 45.00 Hz)
- Filter length: 529 samples (3.306 s)

Used Annotations descriptions: [np.str_('T0'), np.str_('T1'), np.str_('T2')]
Not setting metadata
30 matching events found
No baseline correction applied
0 projection items activated
Using data from preloaded Raw for 30 events and 161 original time points ...
0 bad epochs dropped


[Parallel(n_jobs=1)]: Done  17 tasks      | elapsed:    0.0s
[Parallel(n_jobs=1)]: Done  64 out of  64 | elapsed:    0.1s finished


In [12]:
# Transform dataset into pandas dataframe 
df = pd.DataFrame(features,columns=feature_names)
df['label'] = labels

In [13]:
# Analyze dataframe shape, columns, data types, and sample rows
print(df.info())
display(df.describe())
display(df.head())
print(f" dataframe shape {df.shape}")
df['subject_id'].value_counts()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 420 entries, 0 to 419
Columns: 259 entries, trial_id to label
dtypes: float64(256), int64(2), object(1)
memory usage: 850.0+ KB
None


trial_id     ch1_std      ch1_mean     ch1_max     ch1_min  \
count    420.000000  420.000000  4.200000e+02  420.000000  420.000000   
mean    9537.714286    0.000039 -4.448134e-08    0.000109   -0.000085   
std     5699.931291    0.000020  8.439925e-06    0.000064    0.000041   
min        0.000000    0.000011 -6.903448e-05    0.000020   -0.000301   
25%     4592.000000    0.000025 -3.685732e-06    0.000064   -0.000103   
50%     9568.000000    0.000036 -3.531032e-08    0.000096   -0.000076   
75%    14432.000000    0.000048  3.470348e-06    0.000135   -0.000056   
max    19264.000000    0.000140  3.992565e-05    0.000430   -0.000023   

          ch2_std      ch2_mean     ch2_max     ch2_min     ch3_std  ...  \
count  420.000000  4.200000e+02  420.000000  420.000000  420.000000  ...   
mean     0.000035 -2.560149e-08    0.000095   -0.000075    0.000035  ...   
std      0.000017  6.784997e-06    0.000053    0.000037    0.000017  ...   
min      0.000010 -2.469015e-05    0.000020   -0.000271    0.000011  ...   
25%      0.000021 -3.586249e-06    0.000053   -0.000094    0.000021  ...   
50%      0.000031  1.376996e-07    0.000084   -0.000065    0.000033  ...   
75%      0.000044  3.161087e-06    0.000122   -0.000047    0.000045  ...   
max      0.000125  3.640876e-05    0.000324   -0.000017    0.000112  ...   

         ch62_min    ch63_std     ch63_mean    ch63_max    ch63_min  \
count  420.000000  420.000000  4.200000e+02  420.000000  420.000000   
mean    -0.000064    0.000027  3.501155e-07    0.000077   -0.000063   
std      0.000030    0.000012  5.929466e-06    0.000041    0.000031   
min     -0.000207    0.000009 -4.070357e-05    0.000021   -0.000212   
25%     -0.000085    0.000017 -1.950490e-06    0.000046   -0.000082   
50%     -0.000058    0.000025  4.503886e-07    0.000065   -0.000056   
75%     -0.000039    0.000036  2.750875e-06    0.000101   -0.000038   
max     -0.000020    0.000066  3.207527e-05    0.000192   -0.000018   

         ch64_std     ch64_mean    ch64_max    ch64_min       label  
count  420.000000  4.200000e+02  420.000000  420.000000  420.000000  
mean     0.000025  1.382000e-07    0.000086   -0.000056    1.745238  
std      0.000011  4.736894e-06    0.000049    0.000026    0.825808  
min      0.000008 -4.666118e-05    0.000020   -0.000160    1.000000  
25%      0.000016 -1.988964e-06    0.000044   -0.000073    1.000000  
50%      0.000022  2.129733e-07    0.000069   -0.000051    1.500000  
75%      0.000034  2.059518e-06    0.000125   -0.000037    2.000000  
max      0.000065  2.331532e-05    0.000217   -0.000016    3.000000  

[8 rows x 258 columns]

trial_id subject_id   ch1_std  ch1_mean   ch1_max   ch1_min   ch2_std  \
0         0    S012R04  0.000028 -0.000010  0.000059 -0.000111  0.000028   
1       656    S012R04  0.000033  0.000006  0.000111 -0.000079  0.000022   
2      1312    S012R04  0.000026 -0.000003  0.000058 -0.000077  0.000018   
3      1968    S012R04  0.000026 -0.000004  0.000056 -0.000079  0.000027   
4      2624    S012R04  0.000030  0.000005  0.000105 -0.000056  0.000024   

   ch2_mean   ch2_max   ch2_min  ...  ch62_min  ch63_std     ch63_mean  \
0 -0.000006  0.000062 -0.000093  ... -0.000063  0.000021 -4.080704e-06   
1  0.000003  0.000057 -0.000047  ... -0.000036  0.000020  3.332963e-06   
2 -0.000004  0.000049 -0.000052  ... -0.000047  0.000018 -1.986943e-06   
3 -0.000002  0.000063 -0.000068  ... -0.000062  0.000024  1.415632e-06   
4  0.000003  0.000066 -0.000046  ... -0.000055  0.000021  4.673331e-07   

   ch63_max  ch63_min  ch64_std     ch64_mean  ch64_max  ch64_min  label  
0  0.000067 -0.000056  0.000019 -6.626980e-07  0.000056 -0.000053      1  
1  0.000045 -0.000043  0.000019  3.642887e-06  0.000047 -0.000040      2  
2  0.000036 -0.000043  0.000017 -2.827891e-06  0.000037 -0.000043      1  
3  0.000064 -0.000056  0.000025  2.380007e-06  0.000062 -0.000058      3  
4  0.000047 -0.000053  0.000018  8.674011e-07  0.000049 -0.000041      1  

[5 rows x 259 columns]

 dataframe shape (420, 259)


subject_id
S012R04    30
S010R04    30
S014R04    30
S006R04    30
S004R04    30
S008R04    30
S013R04    30
S001R04    30
S003R04    30
S011R04    30
S015R04    30
S007R04    30
S005R04    30
S009R04    30
Name: count, dtype: int64

In [15]:
#Save dataframe to csv in data folder
df.to_csv('../data/interim/eeg_motor_imagery.csv',index=False)


# Key Takeaways

**Dimension**
- 30 rows per subject X 259 columns
- Features are all float

**Column Naming Convention**
- Features follow the pattern: `ch{channel_number}_{statistic}`
- Example: `ch1_std`, `ch1_mean`, `ch1_max`, `ch1_min`

**Key Identifiers**
- `subject_id`: File name that uniquely identifies the trial subject
- `trial_id`: Timestamp indicating when the trial occurred

**Note:**
Dataset is shorter than wide, more trials should be added. Update: more files have been added to balance the dataset